### DISCUSSION TITLE
Misses references to constants?

### DISCUSSION THREAD

**Discussion by Gioxi** [TNG/ArchUnit#1197] [General]
Hi guys,
thanks for working on this so useful tool, it's great.
I feel like ArchUnit misses dependences on things that are not methods, or at least dependencies on constants (final static) fields.
Maybe you're interested to have a look on that.
Just define a final static member in a class A in package a and a class B in package b that depends on static field of class A.
Then define a rule telling classes in package b shoould not depend on classes in package a, and ArchUnit will just see ... nothing

Thanks again for your super work.

Best,
Giorgio

**Comment by hankem** [comment]
`static final String`s or primitive constants are not referenced as dependencies in the byte code, but just inlined in the constant pool of the class that uses it, cf. https://github.com/TNG/ArchUnit/issues/298#issuecomment-577038056, which is why ArchUnit unfortunately currently cannot detect them. But as these are actually no runtime dependencies, cf. https://github.com/TNG/ArchUnit/issues/298#issuecomment-577535142, it might be okayish to not have a violation for them?

**Reply by Gioxi** [reply]
Thanks for the quick reply.
I understand the technical difficulty.
Concerning the dependency problem, there is one even if at runtime the static final is inlined. Thus ideally I'd expect that it's detected so that I can remove it but, of course that doesn't seem possible.
I understand that a static final dependency can be detected only analyzing source code instead of byte code.

**Reply by hankem** [reply]
> Concerning the dependency problem, there is one even if at runtime the static final is inlined.

Is it? In my experiment of https://github.com/TNG/ArchUnit/issues/298#issuecomment-577535142, the constant was inlined at compile time and actually not needed at runtime.

> ideally I'd expect that it's detected [...] but, of course that doesn't seem possible.
> I understand that a static final dependency can be detected only analyzing source code instead of byte code.

The dependency is actually present in the byte code, see https://github.com/TNG/ArchUnit/issues/298#issuecomment-577040541, but  ASM does not support a direct access to the constant pool, see #309. If somebody finds a smart solution to import these dependencies (ideally without affecting the performance too much 😁), ArchUnit could be extended.
